ezmlm-tstdig.1
.\" $Id: ezmlm-tstdig.1 170 2005-01-04 16:08:10Z bruce $
.TH ezmlm-tstdig 1
.SH NAME
ezmlm-tstdig \- Tests if a digest should be created
.SH SYNOPSIS
.B ezmlm-tstdig
[
.B -k
.I kbytes
] [
.B -m
.I msg
] [
.B \-t
.I time
]
.I dir

.B ezmlm-tstdig
reads files in the list directory
.I dir
and determines if any of the criteria specified by the options are met.
If they are, or if all options specified are 0,
.B ezmlm-tstdig
exits with success (0). Otherwise,
.B ezmlm-tstdig
exits (99), or in case of errors, (100) for permanent and (111) for temporary
errors.

The normal use of
.B ezmlm-tstdig
is to place it in a script so that
.B ezmlm-get(1)
is executed if
.B ezmlm-tstdig
exists 0, but skipped otherwise.

.B ezmlm-tstdig
uses LOCAL to adapt its behavior so that it
can be used from the command line, a script, or from within
.IR dir\fB/editor ,
or
.IR dir\fB/manager .
If invoked from
.IR dir\fB/editor ,
the
.B ezmlm-tstdig
line should end in '|| exit 99'.
When used here,
.B ezmlm-tstdig
if digest generation criteria are met
tests a
timestamp in
.IR dir\fB/tstdig .
If
.I dir\fB/tstdig
is more than 1 hour old or a digest has been successfully created since
.I dir\fB/tstdig
was written,
.B ezmlm-tstdig
will set the timestamp to the current time and exit 0. Otherwise it will
exit 99. This is to guard against initiating duplicate digests when messages
arrive while digesting is in progress.
.SH OPTIONS
.TP
.B \-k
Success if more than
.I kbytes
of message body has accumulated since the latest digest.
.TP
.B \-m
Success if more than
.I msg
messages have accumulated since the latest digest.
.TP
.B \-t
Success if more than
.I time
hours have passed since the latest digest.
.SH USAGE
This is an example script for ezmlm-tstdig usage.

.RS
.nf
#!/bin/sh
#usage: script dir diglist@host
#digest if > 64 kbytes message bodies,
#          > 30 messages, or
#          > 48 hours since last digest.

# ezmlm bin path - modify for your setup
EZPATH='/usr/local/bin/ezmlm'

EZTST="${EZPATH}/ezmlm-tstdig"
EZGET="${EZPATH}/ezmlm-get"
${EZTST} -k64 -m30 -t48 "$1" || exit 0
${EZGET} -t "'$2'" "'$1'" < /dev/null
# improve by testing exit status
.fi
.RE
.PP
.SH BUGS
Obviously,
.B ezmlm-tstdig
should be integrated into the digesting program so that 
.I dir\fB/tstdig
will not be not necessary. This may be done in the next version.
.SH "SEE ALSO"
ezmlm-get(1),
ezmlm-make(1),
ezmlm-manage(1),
ezmlm-send(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

- Added <#C#>, <#T#>, and <#X#> substitutions for (un)subscribe
  confirmation emails, replaced with the hash code, time stamp, and
  action respectively.

- Added option to ezmlm-[un]sub to use a tag other than "manual" in Log.

- Added option to ezmlm-manage to show what response is being sent.

- Fixed behavior of ezmlm-manage -Q flag to match man page.

- Fixed ezmlm-import failing to flush output to the last message,
  and enhanced it to allow reading the mbox from stdin.
  Thanks to Tullio Andreatta.

- Changed to UTF-8

ezmlm-idx-7.1.1, 2010-11-18
===========================

- Fixed bug in getln2 function that prevented error handling.

- Fixed bug in ezmlm-cron that prevented parsing of spaces.

- Fixed unclosed file descriptor on error path in ezmlm-cgi.

ezmlm-idx-7.1.0, 2010-11-03
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Added support for custom subject lines to ezmlm-get (for digests).

- Added a new "omitbottom" control file, equivalent to the -B option for
  both ezmlm-get and ezmlm-manage.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" makefile target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

- Fixed handling Received: headers with the date stamp on a separate line.

- Fixed bug in command-line option processing which caused sender-
  confirmed posting to fail.

- Fixed handling BATV sender addresses with upper-case hex tags.
  Thanks to Kyle Wheeler for pointing this out.

- Adjusted filename in digest attachments to fix problem with Outlook.
  Thanks to Glen Stewart

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been

ezmlm-tstdig.c
/*$Id: ezmlm-tstdig.c 421 2005-09-14 05:46:15Z bruce $*/
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <unistd.h>
#include "stralloc.h"
#include "substdio.h"
#include "readwrite.h"
#include "strerr.h"
#include "sig.h"
#include "getconf.h"
#include "env.h"
#include "fmt.h"
#include "now.h"
#include "lock.h"
#include "sgetopt.h"
#include "errtxt.h"
#include "scan.h"
#include "case.h"
#include "str.h"
#include "open.h"
#include "die.h"
#include "idx.h"
#include "config.h"
#include "auto_version.h"

const char FATAL[] = "ezmlm-tstdig: fatal: ";
const char USAGE[] =
"ezmlm-tstdig: usage: ezmlm-tstdig [-k kbytes] [-m messages] [-t hours] dir";

stralloc line = {0};

substdio ssnum;
char numbuf[16];

char strnum[FMT_ULONG];

int flaglocal = 0;

void main(int argc,char **argv)
{
  char *dir;
  char *local;
  char *def;
  int opt;
  unsigned int pos;
  unsigned long num, digsize, dignum;
  unsigned long cumsize = 0L;
  unsigned long deltanum = 0L;
  unsigned long deltawhen = 0L;
  unsigned long deltasize = 0L;
  unsigned long when, tsttime, digwhen;
  int fd,fdlock;

  (void) umask(022);
  sig_pipeignore();
  when = (unsigned long) now();

  while ((opt = getopt(argc,argv,"k:t:m:vV")) != opteof)
    switch(opt) {
      case 'k':
                if (optarg)
                  scan_ulong(optarg,&deltasize);
                break;
      case 't':
                if (optarg)	/* hours */
                  scan_ulong(optarg,&deltawhen);
                break;
      case 'm':
                if (optarg)
                  scan_ulong(optarg,&deltanum);
                break;
      case 'v':
      case 'V': strerr_die2x(0,"ezmlm-tstdig version: ",auto_version);
      default:
	die_usage();
    }


  startup(dir = argv[optind++]);

  if (argv[optind])
    die_usage();	/* avoid common error of putting options after dir */
  if (!getconf_line(&line,"num",0,dir))
    _exit(99);		/* no msgs no shirt -> no digest */
  if(!stralloc_0(&line)) die_nomem();
  pos = scan_ulong(line.s,&num);
  if (line.s[pos] == ':')
    scan_ulong(line.s+pos+1,&cumsize);

  if (getconf_line(&line,"dignum",0,dir)) {
    if(!stralloc_0(&line)) die_nomem();
    pos = scan_ulong(line.s,&dignum);
    if (line.s[pos] == ':')
      pos += 1 + scan_ulong(line.s+pos+1,&digsize);
    if (line.s[pos] == ':')
      scan_ulong(line.s+pos+1,&digwhen);
  } else {
    dignum = 0L;	/* no file, not done any digest */
    digsize = 0L;	/* nothing digested */
    digwhen = 0L;	/* will force a digest, but the last one was eons  */
			/* ago. ezmlm-get sends it out only if there are   */
			/* messages. This is as it should for new lists.   */
  }
  local = env_get("LOCAL");
  if (local && *local) {			/* in editor or manager */
    def = env_get("DEFAULT");
    if (def && *def) {				/* in manager */
      if (!case_starts(def,"dig") || case_starts(def,"digest-"))
        _exit(0);
    } else {					/* in editor */
      flaglocal = 1;
    }
  }

  if (!deltawhen && !deltasize && !deltanum) _exit(0);
  if ((deltawhen && ((digwhen + deltawhen * 3600L) <= when)) ||
      (deltasize && ((digsize + (deltasize << 2)) <= cumsize)) ||
      (deltanum && ((dignum + deltanum) <= num))) {	/* digest! */
    if (flaglocal) {	/* avoid multiple digests. Of course, ezmlm-tstdig*/
			/* belongs in ezmlm-digest, but it's too late ....*/
      fdlock = lockfile("lock");
      getconf_line(&line,"tstdig",0,dir);
      if (!stralloc_0(&line)) die_nomem();
      scan_ulong(line.s,&tsttime);	/* give digest 1 h to complete */
					/* nobody does digests more often */
      if ((tsttime + 3600L < when) || (tsttime <= digwhen)) {
        fd = open_trunc("tstdign");
        if (fd == -1)
          strerr_die6sys(111,FATAL,ERR_CREATE,dir,"/","tstdign",": ");
        substdio_fdbuf(&ssnum,write,fd,numbuf,sizeof(numbuf));
        if (substdio_put(&ssnum,strnum,fmt_ulong(strnum,when)) == -1)
          strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/","tstdign",": ");
        if (substdio_puts(&ssnum,"\n") == -1)
          strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/","tstdign",": ");
        if (substdio_flush(&ssnum) == -1)
          strerr_die6sys(111,FATAL,ERR_FLUSH,dir,"/","tstdign",": ");
        if (fsync(fd) == -1)
          strerr_die6sys(111,FATAL,ERR_SYNC,dir,"/","tstdign",": ");
        if (close(fd) == -1)
          strerr_die6sys(111,FATAL,ERR_CLOSE,dir,"/","tstdign",": ");
        if (rename("tstdign","tstdig") == -1)
          strerr_die4sys(111,FATAL,ERR_MOVE,"tstdign",": ");
        _exit(0);
      }
    } else
        _exit(0);
  }
  _exit(99);
}

